### PR TITLE
Add support for video stream context

### PR DESCRIPTION
**Pull request by matthewlane**
## Type of change
- Feature

## Description of change
This pull request adds support for video-enabled adapters to participate on both instream and outstream impressions by removing the `video-outstream` mediaType and introducing a new ad unit parameter `mediaTypes` that can be used to configure `video` ad units to either `instream` (default) or `outstream`. No changes are required for adapters that support the current `video` mediaType, though the `mediaTypes` parameter of an adUnit, which may contain a stream context, is now given to adapters in the `callBids` parameter, and adapters may use this to target instream and outstream impressions differently if they wish.

In the case of instream context, bid response objects must contain a `vastUrl` property (this is an existing requirement and not introduced by this PR). Outstream bid response objects must have an associated renderer, either directly on the bid or one defined on the ad unit with the `renderer` ad unit parameter.

## Example AdUnits with Video Context
```JavaScript
pbjs.addAdUnits({
  code: 'video-example',
  sizes: [640, 480],
  mediaTypes: {
    video: { context: 'instream' }
  },

  bids: [{
    bidder: 'appnexusAst',
    params: { placementId: '9333431' }
  }]
});
```

```JavaScript
pbjs.addAdUnits({
  code: 'video-example',
  sizes: [ 640, 480 ],
  mediaTypes: {
    video: { context: 'outstream' }
  },

  // appnexusAst contains a renderer so this isn't required, but if
  // present it will be used in favor of the response renderer
  renderer: {
    url: 'http://cdn.adnxs.com/renderer/video/ANOutstreamVideo.js',
    render: function(bid) {
      ANOutstreamVideo.renderAd({
        targetId: bid.adUnitCode,
        adResponse: bid.adResponse
      });
    }
  },

  bids: [{
    bidder: 'appnexusAst',
    params: { placementId: '5768085' }
  }]
});
```

## Other informatin
@unrulydevelopers unruly adapter is updated by this PR to check for outstream context, please feel free to review and make updates or suggestions to this change

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -696,3 +696,45 @@ export function deepAccess(obj, path) {
...
+ * @property {Object} native native configuration
+ * @property {Object} video video configuration
+ */
+
+/**
+ * Validates an adunit's `mediaTypes` parameter
+ * @param{MediaTypes} mediaTypes mediaTypes parameter to validate
+ * @returns{boolean} If object is valie
+ */
+export function mediaTypesValid(mediaTypes) {
```
Comment (jaiminpanchal27) [inline src/utils.js]: For all functions returning boolean, naming should start with `is` or `has`.  I was going through it over here https://google.github.io/styleguide/jsguide.html#naming-method-names. 
So this function can be renamed as `hasValidMediaType` or `isValidMediaType`

File: src/video.js [hunk 2]
```diff
@@ -1,8 +1,35 @@
...
+export const videoAdUnit = adUnit => adUnit.mediaType === VIDEO_MEDIA_TYPE;
 const nonVideoBidder = bid => !videoAdapters.includes(bid.bidder);
-export const hasNonVideoBidder = adUnit => adUnit.bids.filter(nonVideoBidder).length;
+export const hasNonVideoBidder = adUnit =>
+  adUnit.bids.filter(nonVideoBidder).length;
+
+/*
+ * Validate that the assets required for video context are present on the bid
+ */
+export function videoBidIsValid(bid) {
```
Comment (jaiminpanchal27) [inline src/video.js]: Comment made above for function `mediaTypesValid` applies here too.
Also add missing JSDoc for param and return value and unit test for this function

File: src/video.js [hunk 3]
```diff
@@ -1,8 +1,35 @@
...
+  const bidRequest = getBidRequest(bid.adId);
+  const context =
+    bidRequest && deepAccess(bidRequest, 'mediaTypes.video.context');
+
+  // if context not defined assume default 'instream' for video bids
+  if (!bidRequest || !context || context === INSTREAM) {
+    return !!(bid.vastUrl || bid.vastPayload);
```
Comment (jaiminpanchal27) [inline src/video.js]: We are removing`bid.descriptionUrl` when prebid-cache is live or stable.
So on that note, will this also include `bid.descriptionUrl` ? Not very sure though.

Comment (matthewlane) [reply src/video.js]: It won't, just `bid.vastUrl` or a vast payload is required. `vastPayload` was renamed `vastXml`, will push an updated to fix that here

File: src/adaptermanager.js [hunk 4]
```diff
@@ -48,10 +48,21 @@ function getBids({bidderCode, requestId, bidderRequestId, adUnits}) {
           });
         }
 
+        if (adUnit.mediaTypes) {
+          if (utils.isValidMediaTypes(adUnit.mediaTypes)) {
+            bid = Object.assign({}, bid, { mediaTypes: adUnit.mediaTypes });
+          } else {
+            utils.logError(`mediaTypes is not correctly configured`);
```
Comment (mkendall07) [inline src/adaptermanager.js]: can you add the adUnit.code here so they know which adUnit failed validation?

File: test/spec/video_spec.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    const valid = isValidVideoBid({
+      renderer: {
+        url: 'render.url',
+        render: () => true,
+      }
+    });
+
+    expect(valid).to.be(true);
+
+    utils.getBidRequest.restore();
```
Comment (mkendall07) [inline test/spec/video_spec.js]: these calls are better in a `afterEach` block so failures don't cascade.

**Pull request merged**